### ISSUE TITLE
Support additional OIDC configuration with shoot-oidc-service extension for managed Kyma clusters

### ISSUE THREAD

**Issue by pbochynski** [kyma-project/kyma#18305]
**Description**
Enable the option to trust additional identity provider compliant with OpenID. The provider can be registered in the Kyma cluster and kubernetes API server will authenticate tokens that match the provider issuer.
The complete solution should allow to establish the trust during the provisioning so the cluster can be accessed from fully automated processes (without user presence). To accomplish that the following changes are required:
- enable `shoot-oidc-service` extension for shoot clusters
- add possibility to pass the `OpenIDConnect` resource in the shoot cluster and auxiliary role and role binding as Kyma instance parameter (KEB)

**Reasons**
Many users request a possibility to deploy software to freshly creaqted Kyma clusters in automated way. Changing the default IDP for the cluster is the only solution available for now, but then IDP has to support both human users and service accounts what is usually challenging. With additional OIDC provider it can be used only for system to system authorization and will be much easier to set up.

**Links**
- https://github.com/gardener/oidc-webhook-authenticator/blob/master/README.md
- https://blogs.sap.com/2022/09/23/using-github-actions-openid-connect-in-kubernetes/
- https://github.com/gardener/gardener-extension-shoot-oidc-service/blob/master/docs/usage/openidconnects.md
-

**Comment by maximilianbraun** [comment]
Would love to test it at earliest convinience.

**Comment by kwiatekus** [comment]
The goal would be to allow kyma users to deploy apps (and run tests ) on the provisioned kyma runtimes in the CI jobs . 

Execution can be divided into several stages:
 - [ ] POC - manually provision gardener shoot cluster with [oidc service enabled](https://github.com/gardener/gardener-extension-shoot-oidc-service/blob/master/docs/usage/openidconnects.md) and github configured as an oidc issuer. Try to deploy sample manifest from [gh action ](https://blogs.sap.com/2022/09/23/using-github-actions-openid-connect-in-kubernetes/)
 - [ ] POC - investigate how to run btp CLI commands from gh action as a technical user ( bot ) 
 - [ ] Design the configuration API at the KEB
 - [ ] Design contract between KEB and infrastructure manager (who is in the end creating gardener shoot cluster on behalf of user)
 - [ ] Implement changes in infrastructure manager and KEB

**Comment by kwiatekus** [comment]
OIDC configuration will be included in the gardener provisioning in q2
https://github.com/kyma-project/control-plane/issues/3316#issuecomment-1963822606

**Comment by kwiatekus** [comment]
Kyma Infrastructure Manager (KIM) should apply Operator-Facing OIDC separately from User-facing OIDC(s)

![separateOIDC drawio](https://github.com/kyma-project/kyma/assets/6783567/91506a24-27de-4f04-9f0d-992f019c3d5d)

**Comment by Disper** [comment]
Regarding migration from existing shoots to `Runtime CRs`.
- `oidcConfig` <- Populate with`oidcConfig`
- `additionalOidcConfigs` <-  Populate as the first entry with`oidcConfig`

**Comment by burkardendres** [comment]
Would be great to have this available soon, as a dependency we want to use requires that and also to bring our kya into production it would be kind of a prerequisite.

**Comment by tobiscr** [comment]
Behaviour on KIm side is summarised here: https://github.com/kyma-project/control-plane/issues/3316#issuecomment-2134507015

**Comment by tobiscr** [comment]
To unblock customers who are waiting for the enablement of the OIDC extension in GArdener, we updated the Provisioner last week to activate this extension per default for all new created clusters. So, any new created SKR cluster will have the OIDC extension now enabled and customers can configure their own OIDC provider by creating the corresponding CR in their SKR clusters.

It is planned to start the replacement of the Provisioner with KIM (Kyma Infrasturcture Manager) by end of this month. KIM will also per default enable the OIDC extension for all managed clusters.

**Comment by ptesny** [comment]
> It will be super helpful if we have this capability so that we can bootstrap ArgoCD in the provisioned Kyma cluster so that ArgoCD to take over the rest of the deployments that are coming in from the service pipelines.

That's sth which is already done via a terraform based automation module: https://github.com/quovadis-btp/btp-automation/tree/main/btp-context/runtime-context
At the end of kyma env provisioning and existing ArgoCD instance is being bootstrapped with the newly created kyma cluster. I also create a dedicated ArgoCD project on-the-fly

I'm trying hard to finalize a public root module which will then allow you to use the above child module. Will keep you posted. Thx

**Comment by abhijith-darshan** [comment]
This is great news that the shoot-oidc-extension is available in all newly created kyma clusters. What is the timeline to roll this out to already existing SAP Kyma clusters?

**Comment by PK85** [comment]
Meeting with Benjamin about OIDC migration and release on production (exposed by KEB)
1) Who creates a script copying customer OIDC from main OIDC into additionalOIDC? We need that before doing next steps @tobiscr @ebensom 
2) Does KEB for new cluster already saves input OIDC in main and additionalOIDC sections? @piotrmiskiewicz

**Comment by tobiscr** [comment]
Thanks for your request, here our feedback:

1. We can support here, but we were actually thinking that the migration from `oidcConfig` to `additionalOIDC` will become part of the rollout of the Kyma Operator OIDC configuration (on-call) managed by SRE. Please let us know if you would need support from us.
2. Right now not, it will use the default values configured in KIM (see https://github.com/kyma-project/infrastructure-manager/blob/d8458fca36ee3bf06034f1f531f6c839c348cdef/pkg/gardener/shoot/extender/oidc.go#L13-L34).

**Comment by pbochynski** [comment]
Please consider if we need the OIDC provider for operator access enabled all the time. Some certification programs require that customers grant operator access. Therefore we should consider having the operator OIDC provider configured on demand or at least role binding should be created on demand by the customer that wants the operator to troubleshoot the cluster.

**Comment by tobiscr** [comment]
@ebensom / @PK85 : as @pbochynski mentioned, I fear we have to revisit our decision who we deal with operator access to clusters. Our last agreement to have always access would, at least for PCI relevant runtimes, introduce several disadvantages. the PCI auditor recommended that our ops-team should not be able to access any machines and require to get access granted by customers.

@pbochynski  - do you think we should also revisit our general approach how we manage PCI relevant systems in KLM/KIM? Maybe we should for such systems disable the reconciliation and customers have to enable it actively to receive latest updates.

**Comment by abhijith-darshan** [comment]
Since Kubernetes 1.29 structured authentication gate is enabled by default and Gardener will disable OIDC config from >=1.32 maybe this can be optional setting by default.

It is anyway recommended by Gardener to use structured authentication if the need is less than 64 issuers. Plus there is no need of APIServer restart (but gardener needs shoot reconciliation for the updates to be propagated)

Any plans of allowing multiple issuers for `Kyma` in `BTP cockpit`?

**Comment by tobiscr** [comment]
Hi @abhijith-darshan , thanks for your request! We are going to support multiple OIDC providers soon. The implementation is mostly done, but we have still several migration tasks ahead and our SRE team is working on them.

It's also planned to add [support for multiple OIDC providers](https://github.com/kyma-project/busola/issues/3651) to the Kyma Dashboard.

This feature will still use Gardeners OIDC extension. We move internally to structured authentication, but this is approach is at the moment not used for OIDC configurations provided by customers.

**Comment by abhijith-darshan** [comment]
> Hi [@abhijith-darshan](https://github.com/abhijith-darshan) , thanks for your request! We are going to support multiple OIDC providers soon. The implementation is mostly done, but we have still several migration tasks ahead and our SRE team is working on them.
> 
> It's also planned to add [support for multiple OIDC providers](https://github.com/kyma-project/busola/issues/3651) to the Kyma Dashboard.
> 
> This feature will still use Gardeners OIDC extension. We move internally to structured authentication, but this is approach is at the moment not used for OIDC configurations provided by customers.

Hey @tobiscr thanks for the quick response. Sounds good. Not to de-rail current plans but i would also point out that there might be limitations on the OIDC extension when compared directly to structured authentication.

For example:

- CEL expressions for `claimMapping`
- CEL expressions for `claimValidationRules`
- Multiple Audiences on same `jwt` issuers
